### PR TITLE
WI #1594 Display error when character outside editable range is detected

### PR DESCRIPTION
**Pull request by LANOYDO**

### CODE REVIEW

File: TypeCobol.Test/Parser/Programs/Cobol85/ContinuationLine/ContinuationLine.rdzMix.txt [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+Line 10[19,21] <27, Error, Syntax> - Syntax error : Symbol foo is not referenced
+      -        to foo
+      
+
+      *    OK
+           display "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
+      -            " er zer "
+           display 'endxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx1234xxxxxxx
+      -            ' gggg '
+           display 'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
```
Comment (smedilol) [inline TypeCobol.Test/Parser/Programs/Cobol85/ContinuationLine/ContinuationLine.rdzMix.txt]: Can you add the same display statement but with delimiter `"` ?

**Comment by smedilol** [comment]
Apart from my remarks, I noticed that our incremental parsing report an error most of the time 
with this code:
```cobol
       IDENTIFICATION DIVISION.
       PROGRAM-ID. DVZF0OSM.
       data division.
       working-storage section.
       01 var1 pic X.
       PROCEDURE DIVISION.
           move "azeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee
      -                                                         "afsdsdf
      -         "azazeaez
      -         "azeaz"
                to var1
           .
      
       END PROGRAM DVZF0OSM .
```
Errors "Syntax error : literal not well delimited" and "Syntax error : mismatch input expecting TO" on line:
```cobol
      -                                                         "afsdsdf
```
But this error can dissapear if you save the document.
Code generation never fails.

With the production version of TypeCobol only the second error (mismatch input) is shown.
So I think this can be fixed in another issue.

File: TypeCobol.Test/Parser/Programs/Cobol85/ContinuationLine/ContinuationLine.rdzMix.txt (hunk 1 above)
Comment (LANOYDO) [reply TypeCobol.Test/Parser/Programs/Cobol85/ContinuationLine/ContinuationLine.rdzMix.txt]: done
https://github.com/TypeCobolTeam/TypeCobol/pull/1604/commits/80c7f30d19923763d5f431814c12d34e64324802

File: TypeCobol.Test/Parser/Programs/Cobol85/ContinuationLine/ContinuationLine8.rdzMix.txt [hunk 2]
```diff
@@ -0,0 +1,10 @@
+﻿       IDENTIFICATION division.
+       PROGRAM-ID. MyPGM8.
+       procedure division.
+Line 4[6,6] <1, Error, TextFormat> - Invalid indicator character '[6,6]' : use space for source lines, * or / for comment lines, D or d for debugging lines, - for continuation lines
```
[the remark this answers is not in the document]
Comment (LANOYDO) [reply TypeCobol.Test/Parser/Programs/Cobol85/ContinuationLine/ContinuationLine8.rdzMix.txt]: character * has been modified : now, error on its line is not fired
https://github.com/TypeCobolTeam/TypeCobol/pull/1604/commits/80c7f30d19923763d5f431814c12d34e64324802

File: Codegen/test/resources/input/CobolV6/JsonParse.rdz.cbl [hunk 3]
```diff
@@ -0,0 +1,52 @@
+﻿       IDENTIFICATION DIVISION.
```
[the remark this answers is not in the document]
Comment (LANOYDO) [reply Codegen/test/resources/input/CobolV6/JsonParse.rdz.cbl]: removed
https://github.com/TypeCobolTeam/TypeCobol/pull/1604/commits/80c7f30d19923763d5f431814c12d34e64324802

File: TypeCobol/Compiler/Diagnostics/Resources/DiagnosticMessages.csv [hunk 4]
```diff
@@ -19,7 +19,7 @@ Category;Code;Severity;MessageTemplate;Document;PageNumber;ReferenceText
...
 2;19;2;The number of characters in an hexadecimal alphanumeric literal should be a multiple of 2;1;36;An even number of hexadecimal digits must be specified. Two hexadecimal digits represent one character in a single-byte character set (EBCDIC or ASCII). 
 2;20;2;The number of characters in an hexadecimal national literal should be a multiple of 4;1;41;The number of hexadecimal digits must be a multiple of four. Each group of four hexadecimal digits represents a single national character and must represent a valid code point in UTF-16. 
-2;21;1;The first non blank character of this continuation line must be {0} (continuation of the alphanumeric literal on the previous line);1;55;If the continued line contains an alphanumeric or national literal, the continuation line must contain a hyphen in the indicator area, and the first nonblank character must be a quotation mark
+2;21;1;The delimiter character of this continuation line is different of the initial delimiter of the alphanumeric literal;1;55;If the continued line contains an alphanumeric or national literal, the continuation line must contain the same quotation marks
```
Comment (fm-117) [inline TypeCobol/Compiler/Diagnostics/Resources/DiagnosticMessages.csv]: English : "is different from"

File: TypeCobol/Compiler/Scanner/Scanner.cs [hunk 5]
```diff
@@ -166,6 +166,11 @@ public static void ScanTokensLine(TokensLine tokensLine, MultilineScanState init
             Token nextToken = null;
             while((nextToken = scanner.GetNextToken()) != null)
             {
+                if (nextToken.TokenType == TokenType.AlphanumericLiteral && (!nextToken.HasOpeningDelimiter || !nextToken.HasClosingDelimiter))
+                {
+                    tokensLine.AddDiagnostic(MessageCode.SyntaxErrorInParser,
+                        tokensLine.Indicator.StartIndex, tokensLine.Indicator.EndIndex, "Literal is not well delimited.");
```
Comment (fm-117) [inline TypeCobol/Compiler/Scanner/Scanner.cs]: "is not correctly delimited" ?

File: TypeCobol/Compiler/Scanner/Scanner.cs [hunk 6]
```diff
@@ -329,6 +335,68 @@ public static void ScanTokensLineContinuationGroup(IList<TokensLine> continuatio
                     // Check if the last token so far is an alphanumeric or national literal
                     if (lastTokenOfConcatenatedLineSoFar.TokenFamily == TokenFamily.AlphanumericLiteral)
                     {
+                        if (!lastTokenOfConcatenatedLineSoFar.HasClosingDelimiter)
```
Comment (fm-117) [inline TypeCobol/Compiler/Scanner/Scanner.cs]: You may define some constants for both `'` and `"` to improve readability in this portion of code.

File: TypeCobol/Compiler/Diagnostics/Resources/DiagnosticMessages.csv (hunk 4 above)
Comment (LANOYDO) [reply TypeCobol/Compiler/Diagnostics/Resources/DiagnosticMessages.csv]: done
https://github.com/TypeCobolTeam/TypeCobol/pull/1604/commits/5de0690f865a31192983ee6452e96a892455d316

File: TypeCobol/Compiler/Scanner/Scanner.cs (hunk 5 above)
Comment (LANOYDO) [reply TypeCobol/Compiler/Scanner/Scanner.cs]: done
https://github.com/TypeCobolTeam/TypeCobol/pull/1604/commits/5de0690f865a31192983ee6452e96a892455d316

File: TypeCobol/Compiler/Scanner/Scanner.cs (hunk 6 above)
Comment (LANOYDO) [reply TypeCobol/Compiler/Scanner/Scanner.cs]: done
https://github.com/TypeCobolTeam/TypeCobol/pull/1604/commits/5de0690f865a31192983ee6452e96a892455d316

File: TypeCobol/Compiler/Scanner/Scanner.cs [hunk 7]
```diff
@@ -329,6 +335,70 @@ public static void ScanTokensLineContinuationGroup(IList<TokensLine> continuatio
...
+                                    // no valid closing delimiter
+                                    continuationLine.AddDiagnostic(MessageCode.SyntaxErrorInParser,
+                                        startOfContinuationIndex, startOfContinuationIndex + 1,
+                                        "Closing delimiter of the continuation line is missing.");
+                                    offsetForLiteralContinuation = 0;
+
+                                }
+                                else
+                                {
+                                    // different delimiters
```
Comment (rooksdo) [inline TypeCobol/Compiler/Scanner/Scanner.cs]: Not much but I would add "between start and end delimiters" in comment just to be clearer

File: TypeCobol/Compiler/Scanner/Scanner.cs [hunk 8]
```diff
@@ -329,6 +335,70 @@ public static void ScanTokensLineContinuationGroup(IList<TokensLine> continuatio
                     // Check if the last token so far is an alphanumeric or national literal
                     if (lastTokenOfConcatenatedLineSoFar.TokenFamily == TokenFamily.AlphanumericLiteral)
                     {
+                        if (!lastTokenOfConcatenatedLineSoFar.HasClosingDelimiter)
+                        {
+                            // check delimiters
+                            const char APOSTROPHE = '\'';
```
Comment (smedilol) [inline TypeCobol/Compiler/Scanner/Scanner.cs]: I think it's rather QUOTE and DOUBLE_QUOTE than APOSTROPHE/QUOTATION

File: TypeCobol/Compiler/Scanner/Scanner.cs (hunk 7 above)
Comment (LANOYDO) [reply TypeCobol/Compiler/Scanner/Scanner.cs]: done
https://github.com/TypeCobolTeam/TypeCobol/pull/1604/commits/86f9e27e5ebf7313470d0b3ae58e3e9ea7a080d6

File: TypeCobol/Compiler/Scanner/Scanner.cs (hunk 8 above)
Comment (LANOYDO) [reply TypeCobol/Compiler/Scanner/Scanner.cs]: done
https://github.com/TypeCobolTeam/TypeCobol/pull/1604/commits/86f9e27e5ebf7313470d0b3ae58e3e9ea7a080d6

**Pull request merged**